treeppl/rfc3-functions-vectorization.md

```diff
@@ -0,0 +1,80 @@
+# TreePPL RFC 3: Functions and Vectorization
+
+## Sample function declaration 
+
+	gtr = function(p: Probability[4], r: Rate[5]) => InstantaneousRateMatrix
+	{
+		// ... 
+	}
+
+## Function invocation
+
+Arguments to functions can be labeled (like R, Python...)
+Labelling is optional
+Arguments can have default values
+Reordering the named values should be possible
+mixing of keyword and positional arguments should be possible
+
+Example:
+
+	  Q = gtr(p = pi, r = rep(1, 6))
+	  Q = gtr(pi, rep(1, 6))
+	  Q = gtr(r = rep(1, 6), p = pi)
+	  Q = gtr(p = pi, rep(1, 6))
+	  
```

Comment:
From the documentation:

> In TypeScript, every parameter is assumed to be required by the function. This doesn’t mean that it can’t be given null or undefined, but rather, when the function is called, the compiler will check that the user has provided a value for each parameter. The compiler also assumes that these parameters are the only parameters that will be passed to the function. In short, the number of arguments given to a function has to match the number of parameters the function expects.

Looking at the following example which shows how to specify the default value, I think it makes perfect sense to use `=` (Python uses exactly the same):

    function name(first: string, last = "Newman") {
      return first + " " + last
    }